core/src/main/scala/kafka/network/RequestChannel.scala

```diff
@@ -183,13 +181,16 @@ object RequestChannel extends Logging {
 
   }
 
-  class Response(val request: Request, val responseSend: Option[Send], val responseAction: ResponseAction) {
+  /** responseAsString should only be defined if request logging is enabled */
+  class Response(val request: Request, val responseSend: Option[Send], val responseAction: ResponseAction,
+                 val responseAsString: Option[String]) {
```

Comment:
This is kind of unfortunate but I guess it's a consequence of only having the `Send` at this point. Not seeing any great alternatives, but I guess we could build the string representation into the `Send` object itself. That or maybe we should just keep around the Response object a bit longer when request logging is enabled.

Reply:
We have various Send implementations thought, so it seems like it would be messier than this. Or am I missing something?

I avoided retaining the actual response object as it could be pretty large in some cases. An example is a down converted FetchResponse.

Reply:
There's really only 2-3 implementations of `Send`. What I had in mind was potentially adding the string representation as a member field which was set upon construction and surfaced through `toString`. Might not be any better than this option though, just shifts the ugliness elsewhere.

Reply:
There are 5 implementations although not sure if we'd have to do it for all of them. Also, we only want to compute the `toString` eagerly if request logging is enabled. From that perspective, it seemed best to leave this ugliness in the `RequestChannel` classes instead of spreading them through other classes that shouldn't care about this.

Reply:
Yes, I agree. Just thinking through the other options.

For what it's worth, it seems like the only reason we keep around the request body is to log this message. That means we could do a similar trick and convert it to a string. I guess this is less important because we already have logic to free the records in produce requests though.

Reply:
OK, I'll file a separate JIRA for the request body changes.

Reply:
I filed https://issues.apache.org/jira/browse/KAFKA-5859.